llpc/lower/llpcSpirvLowerTerminator.cpp

```diff
@@ -0,0 +1,159 @@
+ ***********************************************************************************************************************
+ */
+#include "SPIRVInternal.h"
+#include "llpcContext.h"
+#include "llpcDebug.h"
+#include "llpcSpirvLower.h"
+#include "llpcSpirvLowerUtil.h"
+#include "lgc/Builder.h"
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "llpc-spirv-lower-terminator"
+
+using namespace llvm;
+using namespace SPIRV;
+using namespace Llpc;
+
+namespace Llpc {
+
+// =====================================================================================================================
+// Represents the pass of SPIR-V lowering terminators.
+class SpirvLowerTerminator : public SpirvLower, public llvm::InstVisitor<SpirvLowerTerminator> {
+public:
+  SpirvLowerTerminator();
+
+  virtual bool runOnModule(llvm::Module &module);
+  virtual void visitCallInst(llvm::CallInst &callInst);
+
+  static char ID; // ID of this pass
+
+private:
+  SpirvLowerTerminator(const SpirvLowerTerminator &) = delete;
+  SpirvLowerTerminator &operator=(const SpirvLowerTerminator &) = delete;
+
+  // Instructions to be removed; set for tests, vector for order
+  SmallPtrSet<llvm::Instruction *, 8> m_instsForRemoval;
+  SmallVector<llvm::Instruction *, 8> m_removalStack;
+};
+
+// =====================================================================================================================
+// Initializes static members.
+char SpirvLowerTerminator::ID = 0;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of SPIR-V lowering terminator operations
+ModulePass *createSpirvLowerTerminator() {
+  return new SpirvLowerTerminator();
+}
+
+// =====================================================================================================================
+SpirvLowerTerminator::SpirvLowerTerminator() : SpirvLower(ID) {
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+bool SpirvLowerTerminator::runOnModule(Module &module) {
+  LLVM_DEBUG(dbgs() << "Run the pass Spirv-Lower-Terminator\n");
+
+  SpirvLower::init(&module);
+
+  // Kills are only valid in fragment shader model.
+  if (m_shaderStage != ShaderStageFragment)
+    return false;
+
+  // Invoke handling of "kill" instructions.
+  visit(m_module);
+
+  // Remove any dead instructions.
+  bool changed = !m_removalStack.empty();
+  while (!m_removalStack.empty()) {
+    auto deadInst = m_removalStack.pop_back_val();
+    LLVM_DEBUG(dbgs() << "remove: " << *deadInst << "\n");
+    deadInst->dropAllReferences();
+    deadInst->eraseFromParent();
+  }
+  m_instsForRemoval.clear();
```

Comment:
Can we simplify this by only keeping m_instsForRemoval?

I know that it's a set, and we usually have to be careful about iteration order introducing non-determinism. But in this case we're only deleting the instructions, so...?

Reply:
For the second test case I wrote (which is a made up theoretical example), it will fail if the instructions are not removed in reverse order as the second instruction is dependent on the first.  I just assume that was expected behaviour, but perhaps dropAllReference should be preventing that?